### PR TITLE
Reduce TestDatabaseServerAutoDisconnect flakiness

### LINKED ISSUE #34003: `TestDatabaseServerAutoDisconnect ` flakiness
#### Link(s) to logs

https://github.com/gravitational/teleport/actions/runs/6671695688/job/18134147750

#### Relevant snippet

```
=== Failed
=== FAIL: lib/srv/db TestDatabaseServerAutoDisconnect (1.82s)
{"caller":"srv/heartbeat.go:290","component":"db:service:beat","level":"debug","message":"Heartbeat exited.","timestamp":"2023-10-27T20:08:42Z"}
{"caller":"backend/buffer.go:280","component":"buffer","level":"debug","message":"Removing watcher unified_resource (0xc0063d3140) via external close.","timestamp":"2023-10-27T20:08:42Z"}
{"caller":"mysql/test.go:183","component":"mysql","level":"debug","message":"Connection done.","name":"mysql","timestamp":"2023-10-27T20:08:42Z"}
{"caller":"srv/heartbeat.go:290","component":"db:service:beat","level":"debug","message":"Heartbeat exited.","timestamp":"2023-10-27T20:08:42Z"}
{"caller":"db/proxyserver.go:240","component":"db:proxy","level":"debug","message":"MySQL proxy exited.","timestamp":"2023-10-27T20:08:42Z"}
{"caller":"backend/buffer.go:283","component":"buffer","level":"debug","message":"Could not find watcher unified_resource.","timestamp":"2023-10-27T20:08:42Z"}
{"caller":"backend/buffer.go:280","component":"buffer","level":"debug","message":"Removing watcher auth (0xc008144300) via external close.","timestamp":"2023-10-27T20:08:42Z"}
{"caller":"backend/buffer.go:283","component":"buffer","level":"debug","message":"Could not find watcher auth.","timestamp":"2023-10-27T20:08:42Z"}
{"caller":"backend/buffer.go:280","component":"buffer","level":"debug","message":"Removing watcher headless_authentication (0xc0068adbc0) via external close.","timestamp":"2023-10-27T20:08:42Z"}
{"caller":"backend/buffer.go:283","component":"buffer","level":"debug","message":"Could not find watcher headless_authentication.","timestamp":"2023-10-27T20:08:42Z"}
{"caller":"services/watcher.go:247","component":"db:service","level":"debug","message":"Attempting to restart watch after waiting 16.464145959s.","resource-kind":[{"kind":"db","load_secrets":false,"name":""}],"timestamp":"2023-10-27T20:08:42Z"}
{"caller":"services/watcher.go:259","component":"db:service","level":"warning","message":"Restart watch on error: rpc error: code = Canceled desc = grpc: the client connection is closing.","resource-kind":[{"kind":"db","load_secrets":false,"name":""}],"timestamp":"2023-10-27T20:08:42Z"}
```

### PR DESCRIPTION
**Pull request by Tener**
The connection monitor is written in a way which assumes small, incremental changes to the current time. Advancing the clock in big steps ~15s made it behave in a way which violated assumptions made by the test. Theoretically this could happen in production, but only if the watchdog thread was paused for comparable amount of time, which cannot occur in a healthy system. The conditions are therefore unrealistic, and furthermore, the only practical impact would be delay of cancellation by a few seconds.

To work around this issue, instead of making one big jump in time, each advance is divided into 100 steps. 

Fixes https://github.com/gravitational/teleport/issues/34003

Tested locally with:
```
stress -timeout 1h go test -race -v -shuffle=on -run TestDatabaseServerAutoDisconnect -count 250 ./lib/srv/db -failfast -timeout 1h
```

### CODE REVIEW

**Review by greedy52 — COMMENTED** [review]
> The connection monitor is written in a way which assumes small, incremental changes to the current time.

I am try to grasp how this works. Do you mind explain a little more or point to the problematic code? Thanks!

In the original CI failure, it seems we already received `events.ClientDisconnectCode` ? I wonder if we also have a race that event is emitted but connection is not closed yet.

**Comment by Tener** [comment]
> I am try to grasp how this works. Do you mind explain a little more or point to the problematic code? Thanks!

Sure thing. I think the key part is these few lines in monitor:

https://github.com/gravitational/teleport/blob/f6cde21081367e7018539190b788f5b6e09fcaff/lib/srv/monitor.go#L341-L361

If the system has high load, and we move the time by large chunks, the "wake up time" (which is discarded element from `idleTime` chan) may be very different from the time taken into calculation in `w.Clock.Since()`. There is a race condition between the clock triggering and calculation of `since` taking place.

But there are also parallel, background threads (two: one for engine, one for proxy) that perform `.Read()` on `TrackingReadConn`:

https://github.com/gravitational/teleport/blob/f6cde21081367e7018539190b788f5b6e09fcaff/lib/srv/monitor.go#L537-L548

This happens asynchronously and will execute after the context is closed; this leads to slightly weird situation where the read fails (`err != nil`), but we still update the activity tracker, which makes the connection appear "active". I was wondering if we should check the err value in that function, but ultimately I decided against making changes without a clear-cut case.

> In the original CI failure, it seems we already received events.ClientDisconnectCode ? I wonder if we also have a race that event is emitted but connection is not closed yet.

I think you are right that there is a race condition here: we emit the event, but the connection is still open for the time being, so we may succeed in using it.

**Comment by Tener** [comment]
@greedy52 https://github.com/gravitational/teleport/pull/34060/commits/c895bcd2cb270738a469390db790086d8c4385e2 reorders the operations: we now emit disconnect event after actually closing the connection. The downside is that we cannot use `w.Context` anymore, but doing so feels like a hack anyway: audit events should be independent from the state of connection.

**Comment by greedy52** [comment]
https://github.com/gravitational/teleport/blob/f6cde21081367e7018539190b788f5b6e09fcaff/lib/srv/monitor.go#L342-L345

Sorry, I am still struggling on this. Is the concern that clock is advanced between `<-idleTime` and `w.Clock.Since(clientLastActive)`? 

Would it help if we have something like `w.Tracker.SinceClientLastActive()` so since is calculated while locking the activity update (maybe not as the raw read already happened?)? 

How would doing small time increments help?

> @greedy52 https://github.com/gravitational/teleport/commit/c895bcd2cb270738a469390db790086d8c4385e2 reorders the operations: we now emit disconnect event after actually closing the connection. The downside is that we cannot use w.Context anymore, but using it feels like a hack anyway: audit events should be independent from the state of connection.

Looking at the history though, it's emitted first on purpose https://github.com/gravitational/teleport/pull/31066. (the monitor's context can be closed right away once the underlying conn is closed, then there is no chance to emit). Also seems kube forwarding is passing the same context for `TrackingReadConn` and `Monitor`. Maybe pass in a separate `EmitterContext`? or maybe a callback when the monitor is closed?

**Comment by Tener** [comment]
> Sorry, I am still struggling on this. Is the concern that clock is advanced between <-idleTime and w.Clock.Since(clientLastActive)?

Exactly this. If you are unlucky 15s will pass in between these two calls.

> Would it help if we have something like w.Tracker.SinceClientLastActive() since is calculated while locking the activity update? (maybe not as the raw read already happened?)

> How would doing small time increments help?

I'm hesitant to adjust working code just to suit a test. The small increments are each followed by a small sleep, which helps us achieve correct behaviour. The existing race is already rather unlikely, but now I'm making it even more unlikely. I'd much rather fix the race entirely, but I don't see how to do it without heavy refactoring. It may well be impossible to trigger in practice, unless go scheduler somehow forgets to run the goroutine for extended periods of time.

> Looking at the history though, it's emitted first on purpose #31066. (the monitor's context can be closed right away once the underlying conn is closed, then there is no chance to emit). Also seems kube forwarding is passing the same context for `TrackingReadConn` and `Monitor`. Maybe pass in a separate `EmitterContext`? or maybe a callback when the monitor is closed?

A separate `EmitterContext` would probably be best. I'll take a look at how complex this would be. In the meanwhile, a fresh, time-limited context is used instead.

**Comment by greedy52** [comment]
> The small increments are each followed by a small sleep, which helps us achieve correct behaviour.

cool. I am good with using the small sleeps to workaround the test, as you stated it's not a real-life issue.

I think we do want fix the other race too. The minor problem with `context.WithTimeout(context.TODO(), time.Second)` is we don't know if `time.Second` is too much or not enough so best if we could find a better way. If that's not something easy, I think one alternative is maybe add a test hook (similar to `OnReconcile`) in db server's `handConnection` to indicate the connection is closed.

**Comment by Tener** [comment]
@greedy52 I've added `EmitterContext` in https://github.com/gravitational/teleport/pull/34060/commits/94f2136de6b720e300bd5c5ea3f006ab4e27e1c4.

**Review by greedy52 — APPROVED** [review]
Thanks! LGTM. The CI may have caught something tho.

File: lib/srv/db/server_test.go [hunk 1]
```diff
@@ -236,6 +237,16 @@ func TestDatabaseServerAutoDisconnect(t *testing.T) {
 	require.NoError(t, pgConn.Close(ctx))
 }
 
+func advanceInSteps(clock clockwork.FakeClock, total time.Duration) {
+	step := total / 100
```
Comment (codingllama) [inline lib/srv/db/server_test.go:241]: Should we make sure step is at least 1?

File: lib/srv/db/server_test.go [hunk 2]
```diff
@@ -236,6 +237,16 @@ func TestDatabaseServerAutoDisconnect(t *testing.T) {
 	require.NoError(t, pgConn.Close(ctx))
 }
 
+func advanceInSteps(clock clockwork.FakeClock, total time.Duration) {
```
Comment (codingllama) [inline lib/srv/db/server_test.go:240]: This one could do with a comment, as generally this isn't something we need to be worried about.

File: lib/srv/monitor.go [hunk 3]
```diff
@@ -224,6 +230,8 @@ type MonitorConfig struct {
 	ServerID string
 	// Emitter is events emitter
 	Emitter apievents.Emitter
+	// EmitterContext is long-lived context suitable to be used with Emitter.
+	EmitterContext context.Context
```
Comment (codingllama) [inline lib/srv/monitor.go:234]: How is this different from the other Context field?

File: lib/srv/monitor.go [hunk 4]
```diff
@@ -71,6 +71,8 @@ type ConnectionMonitorConfig struct {
 	ServerID string
 	// Emitter allows events to be emitted.
 	Emitter apievents.Emitter
+	// EmitterContext is long-lived context suitable to be used with Emitter
+	EmitterContext context.Context
```
Comment (codingllama) [inline lib/srv/monitor.go:75]: Obligatory https://github.com/golang/go/wiki/CodeReviewComments#contexts. (Although, tbh, I'm almost giving up quoting this one.)

File: lib/srv/db/server_test.go
[the remark this answers is not in the document]
Comment (Tener) [reply lib/srv/db/server_test.go:1]: Well, the assumption isn't wrong in normal usage. It is only the fake clock which behaves in this way, and I'm hesitant about changing the production code in significant way to suit the tests.

File: lib/srv/monitor.go (hunk 4 above)
Comment (Tener) [reply lib/srv/monitor.go:75]: We kind of violate this nearly everywhere, though. For audit events, in particular, I think we should always use server exit context or equivalent. Unless the server is shutting down, hard, we should still try to emit the event. 

I'd argue this kind of sharing is fine, too, at least for the purpose of synchronising the lifetimes of different services. Sharing credentials would be much more icky.

File: lib/srv/monitor.go (hunk 3 above)
Comment (Tener) [reply lib/srv/monitor.go:234]: This one is long-lived: if the server is not shutting down, we should always attempt to emit the audit event. The other one is short-lived and may be cancelled when we close the connection, which this file is all about.

Comment (codingllama) [reply lib/srv/monitor.go:234]: Gotcha. I would note that in the godoc.

File: lib/srv/monitor.go (hunk 4 above)
Comment (codingllama) [reply lib/srv/monitor.go:75]: Yep.

File: lib/srv/db/server_test.go
[the remark this answers is not in the document]
Comment (codingllama) [reply lib/srv/db/server_test.go:1]: Fair enough.

File: lib/srv/monitor.go (hunk 4 above)
Comment (Tener) [reply lib/srv/monitor.go:75]: Still, thanks for the reminder, I have actually forgotten about this style issue.

**Comment by Tener** [comment]
Looks like it needs another review still. 

PTAL @smallinsky or @rosstimothy ?

**Review by rosstimothy — COMMENTED** [review]
Couldn't we also rewrite the test to not rely on a fake clock and set the idle timeout to something less than 30s?

**Review by codingllama — APPROVED** [review]
Approved in case you want to push further changes for later. Let us know what the plan is.

**Comment by zmb3** [comment]
Given that this has required approvals I am going to add it to the merge queue since this flakiness is still slowing us down.

**Pull request merged**